pyinteraph/path_analysis.py

```diff
@@ -0,0 +1,577 @@
+import argparse
+import logging as log
+import numpy as np
+import networkx as nx
+import MDAnalysis as mda
+import itertools
+import re
+import matplotlib.pyplot as plt
+import seaborn as sns
+import matplotlib.colors as mplcolors
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+
+def build_graph(fname, pdb = None):
+    """Build a graph from the provided matrix"""
+
+    try:
+        data = np.loadtxt(fname)
+    except:
+        errstr = f"Could not load file {fname} or wrong file format."
+        raise ValueError(errstr)
+    # if the user provided a reference structure
+    if pdb is not None:
+        try:
+            # generate a Universe object from the PDB file
+            u = mda.Universe(pdb)
+        except Exception as e:
+            errstr = \
+                f"Exception caught during creation of the Universe: {e}"
+            raise ValueError(errstr)      
+        # generate identifiers for the nodes of the graph
+        identifiers = [f"{r.segment.segid}{r.resnum}" for r in u.residues]
+    # if the user did not provide a reference structure
+    else:
+        # generate automatic identifiers going from 1 to the
+        # total number of residues considered
+        identifiers = [str(i) for i in range(1, data.shape[0]+1)]
+    
+    # generate a graph from the data loaded
+    G = nx.Graph(data)
+    # set the names of the graph nodes (in place)
+    node_names = dict(zip(range(data.shape[0]), identifiers))
+    nx.relabel_nodes(G, mapping = node_names, copy = False)
+    # return the idenfiers and the graph
+    return identifiers, G
+
+def convert_input_to_list(user_input, identifiers, pdb = False):
+    """Take in a string (e.g. A12:A22,A13... if a PDB file is supplied)
+    and a list of names of all the residues (graph nodes). Replaces the 
+    range indicated by the colon with all resiues in that range and 
+    keeps all residues separated by commas. Removes duplicates. Takes in
+    a string e.g. 1,3,4:56 if no PDB file is supplied.
+    """
+
+    # Check if PDB file is supplied
+    if pdb:
+        # Find all residues separated by commas by
+        # replacing all colon residues with ''
+        input_comma = re.sub('\w+:\w+', '', user_input)
```

Comment:
I used regex because I wasn't sure how else to extract only the characters separated by `:`. but I guess I can try to figure it out. Shouldn't `2:4:6` be considered an error and ignored? If a user types in`2:4:6` what do we want to do?

Reply:
yes, `2:4:6` should be considered an error, but I think it's not with your actual code. One way to do this would be simply to split by `,` then split each resulting element by `:` and see how many are there, and behave appropriately

Reply:
Ahhh, I completely forgot that when you split by colon a second time, it will create a sub list. For some reason I thought it wouldn't and I would have no way to differentiate between colons and commas. It should be fixed now.